src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -4760,6 +4760,24 @@ private void EmitMatchCharacterClass(string charClass, bool caseInsensitive, Loc
                 return;
             }
 
+            // Next if the character class contains nothing but a single Unicode category, we can calle char.GetUnicodeCategory and
+            // compare against it.  It has a fast-lookup path for ASCII, so is as good or better than any lookup we'd generate (plus
+            // we get smaller code), and it's what we'd do for the fallback (which we get to avoid generating) as part of CharInClass.
+            if (!invariant && RegexCharClass.TryGetSingleUnicodeCategory(charClass, out UnicodeCategory category, out bool negated))
```

Comment:
Why `!invariant`? Shouldn't this work regardless?

Reply:
We already called ToLower if !invariant.  For invariant, we build that knowledge into the lookup table.  If we wanted to use this path for invariant, we'd need to call ToLower, and the ToLower+GetUnicodeCategory is more expensive than the lookup table that factors in the invariance.
https://github.com/dotnet/runtime/blob/3203ee2cf51073274d5153d3cb519df3b5d8eb5a/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs#L4722-L4724